keyboards/grid/grid_650/grid_650.c

```diff
@@ -0,0 +1,643 @@
+                            //ble not connected enable blue blink 
+                            set_status_led(BLE_ON,curr_stat);
+                            curr_stat = BLE_ON;                            
+                        }                              
+                    }
+                    break;
+                    default:break;
+                }
+            }            
+        }
+        #endif
+
+
+    }
+    #endif
+
+    #ifdef ENABLE_SLEEP
+    if (has_anykey(keyboard_report) || (PIND & 0b10000000)) {
+    // if (has_anykey(keyboard_report)) {
+        // has key press clr counter
+        sleep_counter = 0;
+        sleep_activated = 0;
+        
+        #ifdef ENABLE_DEEP_SLEEP
+            ble_on();
+        #endif
+
+        // backlight_enable();
+        // breathing_enable();
+    } else if (sleep_counter <= SLEEP_WAIT) {
+        // // always goto IDLE mode to save power
+        
+    } else if (sleep_counter > SLEEP_WAIT) {
+        //when counter is big enough goto POWERDOWN mode
+        // sleep_counter = SLEEP_WAIT;
+        sleep_activated = 1;
+
+        
+        #ifdef ENABLE_DEEP_SLEEP
+            ble_off();
+        #endif
+            // suspend_power_down(); //use qmk provided api not working properly
+
+            // breathing_disable();
+            // backlight_disable();     
+
+            set_status_led(NONE,curr_stat);
+            curr_stat = NONE;    
+
+            // Watchdog Interrupt Mode
+            wdt_intr_enable_copy(WDTO_60MS);
+
+            // cli();
+            
+            // set_sleep_mode(SLEEP_MODE_IDLE);
+            set_sleep_mode(SLEEP_MODE_PWR_DOWN);//power down mode saves better but resets the system
+            sleep_enable();
+            // sleep_bod_disable(); //try to disable bod reset
+            sei();
+            
+            sleep_cpu();
+            sleep_disable();
+
+            // Disable watchdog after sleep
+            wdt_disable();
+            //must have some delay or reset will occur(40ms is ok)
+            wait_ms(10);
+
+        //need to think about how to save energy
+        //can key scan interval be changed?       
+    }
+    #endif
+
+    matrix_scan_user();
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    // put your per-action keyboard code here
+    // runs for every action, just before processing by the firmware
+
+    return process_record_user(keycode, record);
+}
+
+void led_set_user(uint8_t usb_led) {
+  #ifndef BLE_ENABLE
+
+  if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
+    //set white led low
+    writePinLow(B6);
+    //set orange led high
+    writePinHigh(B7);
+  } else {
+    //set orange led low
+    writePinLow(B7);
+    //set white led high
+    writePinHigh(B6);
+  }
+  #endif
+}
+
+void suspend_power_down_kb(void)
+{
+    #ifndef BLE_ENABLE
+    //set white led low
+    writePinLow(B6);
+    //set orange led low
+    writePinLow(B7);
+    #endif
+}
+
+uint8_t init_mcp23018(void) {
```

Comment:
Should use `drivers/gpio/mcp23018.c` for this stuff.

There are a few examples in the repo already of this being used.